File: android-app/src/main/java/fr/cph/chicago/core/activity/BikeStationActivity.java
/**
 * Copyright 2017 Carl-Philipp Harmant
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.activity;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.Toolbar;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.annimon.stream.Stream;

import java.util.List;

import butterknife.BindColor;
import butterknife.BindString;
import butterknife.BindView;
import butterknife.ButterKnife;
import fr.cph.chicago.R;
import fr.cph.chicago.core.listener.GoogleMapDirectionOnClickListener;
import fr.cph.chicago.core.listener.GoogleMapOnClickListener;
import fr.cph.chicago.core.listener.GoogleStreetOnClickListener;
import fr.cph.chicago.data.PreferencesImpl;
import fr.cph.chicago.entity.BikeStation;
import fr.cph.chicago.entity.Position;
import fr.cph.chicago.entity.enumeration.TrainLine;
import fr.cph.chicago.rx.observable.ObservableUtil;
import fr.cph.chicago.rx.observer.BikeAllBikeStationsObserver;
import fr.cph.chicago.util.Util;

/**
 * Activity the list of train stations
 *
 * @author Carl-Philipp Harmant
 * @version 1
 */
@SuppressWarnings("WeakerAccess")
public class BikeStationActivity extends AbstractStationActivity {

    @BindView(R.id.activity_station_swipe_refresh_layout)
    SwipeRefreshLayout swipeRefreshLayout;
    @BindView(R.id.activity_favorite_star)
    ImageView favoritesImage;
    @BindView(R.id.activity_bike_station_streetview_image)
    ImageView streetViewImage;
    @BindView(R.id.activity_bike_station_steetview_text)
    TextView streetViewText;
    @BindView(R.id.activity_map_image)
    ImageView mapImage;
    @BindView(R.id.activity_map_direction)
    ImageView directionImage;
    @BindView(R.id.map_container)
    LinearLayout mapContainer;
    @BindView(R.id.walk_container)
    LinearLayout walkContainer;
    @BindView(R.id.favorites_container)
    LinearLayout favoritesImageContainer;
    @BindView(R.id.activity_bike_station_value)
    TextView bikeStationValue;
    @BindView(R.id.toolbar)
    Toolbar toolbar;
    @BindView(R.id.favorites_bikes_list)
    LinearLayout container;

    @BindString(R.string.bundle_bike_station)
    String bundleBikeStation;
    @BindString(R.string.bike_available_bikes)
    String bikeAvailableBikes;
    @BindString(R.string.bike_available_docks)
    String bikeAvailableDocks;

    @BindColor(R.color.grey_5)
    int grey_5;
    @BindColor(R.color.red)
    int red;
    @BindColor(R.color.green)
    int green;
    @BindColor(R.color.yellowLineDark)
    int yellowLineDark;

    private BikeStation bikeStation;
    private boolean isFavorite;

    @Override
    protected final void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (!this.isFinishing()) {
            setContentView(R.layout.activity_bike_station);
            ButterKnife.bind(this);
            bikeStation = getIntent().getExtras().getParcelable(bundleBikeStation);
            if (bikeStation != null) {
                final double latitude = bikeStation.getLatitude();
                final double longitude = bikeStation.getLongitude();

                swipeRefreshLayout.setOnRefreshListener(
                    () -> ObservableUtil.createAllBikeStationsObservable()
                        .subscribe(new BikeAllBikeStationsObserver(this, bikeStation.getId(), swipeRefreshLayout))
                );

                isFavorite = isFavorite();

                // Call google street api to load image
                loadGoogleStreetImage(new Position(latitude, longitude), streetViewImage, streetViewText);

                mapImage.setColorFilter(grey_5);
                directionImage.setColorFilter(grey_5);

                if (isFavorite) {
                    favoritesImage.setColorFilter(yellowLineDark);
                } else {
                    favoritesImage.setColorFilter(grey_5);
                }

                favoritesImageContainer.setOnClickListener(view -> switchFavorite());
                bikeStationValue.setText(bikeStation.getStAddress1());
                streetViewImage.setOnClickListener(new GoogleStreetOnClickListener(latitude, longitude));
                mapContainer.setOnClickListener(new GoogleMapOnClickListener(latitude, longitude));
                walkContainer.setOnClickListener(new GoogleMapDirectionOnClickListener(latitude, longitude));

                drawData();
            }
            setToolBar();
        }
    }

    private void setToolBar() {
        toolbar.inflateMenu(R.menu.main);
        toolbar.setOnMenuItemClickListener((item -> {
            swipeRefreshLayout.setRefreshing(true);
            ObservableUtil.createAllBikeStationsObservable()
                .subscribe(new BikeAllBikeStationsObserver(BikeStationActivity.this, bikeStation.getId(), swipeRefreshLayout));
            return false;
        }));
        Util.INSTANCE.setWindowsColor(this, toolbar, TrainLine.NA);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            toolbar.setElevation(4);
        }
        toolbar.setTitle(bikeStation.getName());
        toolbar.setNavigationIcon(R.drawable.ic_arrow_back_white_24dp);
        toolbar.setOnClickListener(v -> finish());
    }

    private void drawData() {
        final Context context = getApplicationContext();

        final LinearLayout availableLayout = new LinearLayout(context);
        final LinearLayout availableBikes = new LinearLayout(context);
        final LinearLayout availableDocks = new LinearLayout(context);

        final TextView availableBike = new TextView(context);
        final TextView availableDock = new TextView(context);
        final TextView amountBike = new TextView(context);
        final TextView amountDock = new TextView(context);

        container.removeAllViews();
        container.setOrientation(LinearLayout.HORIZONTAL);
        availableLayout.setOrientation(LinearLayout.VERTICAL);
        availableBikes.setOrientation(LinearLayout.HORIZONTAL);
        availableBike.setText(bikeAvailableBikes);
        availableBike.setTextColor(grey_5);
        availableBikes.addView(availableBike);
        amountBike.setText(String.valueOf(bikeStation.getAvailableBikes()));
        if (bikeStation.getAvailableBikes() == 0) {
            amountBike.setTextColor(red);
        } else {
            amountBike.setTextColor(green);
        }
        availableBikes.addView(amountBike);
        availableLayout.addView(availableBikes);
        availableDocks.setOrientation(LinearLayout.HORIZONTAL);
        availableDock.setText(bikeAvailableDocks);
        availableDock.setTextColor(grey_5);
        availableDocks.addView(availableDock);
        amountDock.setText(String.valueOf(bikeStation.getAvailableDocks()));
        if (bikeStation.getAvailableDocks() == 0) {
            amountDock.setTextColor(red);
        } else {
            amountDock.setTextColor(green);
        }
        availableDocks.addView(amountDock);
        availableLayout.addView(availableDocks);
        container.addView(availableLayout);
    }

    @Override
    public void onRestoreInstanceState(final Bundle savedInstanceState) {
        super.onRestoreInstanceState(savedInstanceState);
        bikeStation = savedInstanceState.getParcelable(bundleBikeStation);
    }

    @Override
    public void onSaveInstanceState(final Bundle savedInstanceState) {
        savedInstanceState.putParcelable(bundleBikeStation, bikeStation);
        super.onSaveInstanceState(savedInstanceState);
    }

    /**
     * Is favorite or not ?
     *
     * @return if the station is favorite
     */
    @Override
    protected boolean isFavorite() {
        final List<String> favorites = PreferencesImpl.INSTANCE.getBikeFavorites(getApplicationContext());
        return Stream.of(favorites)
            .filter(favorite -> Integer.valueOf(favorite) == bikeStation.getId())
            .findFirst()
            .isPresent();
    }

    public void refreshStation(@NonNull final BikeStation station) {
        this.bikeStation = station;
        drawData();
    }

    /**
     * Add/remove favorites
     */
    private void switchFavorite() {
        if (isFavorite) {
            Util.INSTANCE.removeFromBikeFavorites(bikeStation.getId(), swipeRefreshLayout);
            favoritesImage.setColorFilter(grey_5);
            isFavorite = false;
        } else {
            Util.INSTANCE.addToBikeFavorites(bikeStation.getId(), swipeRefreshLayout);
            PreferencesImpl.INSTANCE.addBikeRouteNameMapping(getApplicationContext(), Integer.toString(bikeStation.getId()), bikeStation.getName());
            favoritesImage.setColorFilter(yellowLineDark);
            isFavorite = true;
        }
    }
}

File: android-app/src/main/kotlin/fr/cph/chicago/core/activity/station/BikeStationActivity.kt
/**
 * Copyright 2021 Carl-Philipp Harmant
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.cph.chicago.core.activity.station

import android.os.Bundle
import androidx.appcompat.widget.Toolbar
import com.google.android.material.appbar.MaterialToolbar
import fr.cph.chicago.R
import fr.cph.chicago.core.listener.GoogleStreetOnClickListener
import fr.cph.chicago.core.listener.OpenMapDirectionOnClickListener
import fr.cph.chicago.core.listener.OpenMapOnClickListener
import fr.cph.chicago.core.model.BikeStation
import fr.cph.chicago.core.model.Position
import fr.cph.chicago.databinding.ActivityBikeStationBinding
import fr.cph.chicago.redux.AddBikeFavoriteAction
import fr.cph.chicago.redux.BikeStationAction
import fr.cph.chicago.redux.RemoveBikeFavoriteAction
import fr.cph.chicago.redux.State
import fr.cph.chicago.redux.Status
import fr.cph.chicago.redux.store
import fr.cph.chicago.util.Color
import org.rekotlin.StoreSubscriber
import timber.log.Timber

/**
 * Activity the list of train stations
 *
 * @author Carl-Philipp Harmant
 * @version 1
 */
class BikeStationActivity : StationActivity(), StoreSubscriber<State> {

    private lateinit var bikeStation: BikeStation
    private lateinit var binding: ActivityBikeStationBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityBikeStationBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupView(
            swipeRefreshLayout = binding.activityStationSwipeRefreshLayout,
            streetViewImage = binding.header.streetViewImage,
            streetViewProgressBar = binding.header.streetViewProgressBar,
            streetViewText = binding.header.streetViewText,
            favoritesImage = binding.header.favorites.favoritesImage,
            mapImage = binding.header.favorites.mapImage,
            favoritesImageContainer = binding.header.favorites.favoritesImageContainer
        )

        bikeStation = intent.extras?.getParcelable(getString(R.string.bundle_bike_station))
            ?: BikeStation.buildUnknownStation()
        position = Position(bikeStation.latitude, bikeStation.longitude)

        loadGoogleStreetImage(position)

        handleFavorite()

        binding.bikeStationValue.text = bikeStation.address
        streetViewImage.setOnClickListener(GoogleStreetOnClickListener(position.latitude, position.longitude))
        binding.header.favorites.mapContainer.setOnClickListener(OpenMapOnClickListener(position.latitude, position.longitude))
        binding.header.favorites.walkContainer.setOnClickListener(OpenMapDirectionOnClickListener(position.latitude, position.longitude))

        drawData()
        buildToolbar(binding.included.toolbar)
    }

    override fun onPause() {
        super.onPause()
        store.unsubscribe(this)
    }

    override fun onResume() {
        super.onResume()
        store.subscribe(this)
    }

    override fun newState(state: State) {
        Timber.d("New state")
        when (state.bikeStationsStatus) {
            Status.FAILURE, Status.FULL_FAILURE -> util.showSnackBar(swipeRefreshLayout, state.bikeStationsErrorMessage)
            Status.ADD_FAVORITES -> {
                if (applyFavorite) {
                    util.showSnackBar(swipeRefreshLayout, R.string.message_add_fav, true)
                    applyFavorite = false
                    favoritesImage.setColorFilter(Color.yellowLineDark)
                }
            }
            Status.REMOVE_FAVORITES -> {
                if (applyFavorite) {
                    util.showSnackBar(swipeRefreshLayout, R.string.message_remove_fav, true)
                    applyFavorite = false
                    favoritesImage.drawable.colorFilter = mapImage.drawable.colorFilter
                }
            }
            else -> {
                state.bikeStations
                    .filter { station -> bikeStation.id == station.id }
                    .elementAtOrElse(0) { BikeStation.buildDefaultBikeStationWithName("error") }
                    .also { station ->
                        if (station.name != "error") {
                            refreshStation(station)
                            intent.extras?.putParcelable(getString(R.string.bundle_bike_station), station)
                        } else {
                            Timber.w("Train station id [%s] not found", bikeStation.id)
                            util.showOopsSomethingWentWrong(swipeRefreshLayout)
                        }
                    }
            }
        }
        stopRefreshing()
    }

    override fun refresh() {
        super.refresh()
        store.dispatch(BikeStationAction())
        loadGoogleStreetImage(position)
    }

    override fun buildToolbar(toolbar: Toolbar) {
        super.buildToolbar(toolbar)
        toolbar.title = bikeStation.name
    }

    private fun drawData() {
        if (bikeStation.availableBikes == -1) {
            binding.availableBikes.text = "?"
            binding.availableBikes.setTextColor(Color.orange)
        } else {
            binding.availableBikes.text = util.formatBikesDocksValues(bikeStation.availableBikes)
            val color = if (bikeStation.availableBikes == 0) Color.red else Color.green
            binding.availableBikes.setTextColor(color)
        }
        if (bikeStation.availableDocks == -1) {
            binding.availableDocks.text = "?"
            binding.availableDocks.setTextColor(Color.orange)
        } else {
            binding.availableDocks.text = util.formatBikesDocksValues(bikeStation.availableDocks)
            val color = if (bikeStation.availableDocks == 0) Color.red else Color.green
            binding.availableDocks.setTextColor(color)
        }
    }

    public override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        bikeStation = savedInstanceState.getParcelable(getString(R.string.bundle_bike_station))
            ?: BikeStation.buildUnknownStation()
        position = Position(bikeStation.latitude, bikeStation.longitude)
    }

    public override fun onSaveInstanceState(savedInstanceState: Bundle) {
        if (::bikeStation.isInitialized) savedInstanceState.putParcelable(getString(R.string.bundle_bike_station), bikeStation)
        super.onSaveInstanceState(savedInstanceState)
    }

    /**
     * Is favorite or not ?
     *
     * @return if the bike station is favorite
     */
    override fun isFavorite(): Boolean {
        return preferenceService.isBikeStationFavorite(bikeStation.id)
    }

    private fun refreshStation(station: BikeStation) {
        this.bikeStation = station
        drawData()
    }

    /**
     * Add/remove favorites
     */
    override fun switchFavorite() {
        super.switchFavorite()
        if (isFavorite()) {
            store.dispatch(RemoveBikeFavoriteAction(bikeStation.id))
        } else {
            store.dispatch(AddBikeFavoriteAction(bikeStation.id, bikeStation.name))
        }
    }
}
